proxy: added endpoint refresh and timeout configuration values

webner opened the pull request:
the default dial timeout was set to 30 seconds this made the proxy a pain to use
in failure scenarios.

fixes 2862

xiang90 commented:
@webner I am OK with making these configurable. But probably we should not give a default timeout for `read`. We have long pulling watch request.

webner commented:
good point. I removed the default read timeout (set it to 0) and added comment in the documentation that changing the read timeout will break watches.

xiang90 commented on Documentation/configuration.md:
```diff
@@ -105,6 +105,27 @@ To start etcd automatically using custom settings at startup in Linux, using a [
++ Time (in milliseconds) for a dial to timeout.
++ default: 1000
+
+##### -proxy-write-timeout
```
can we also comment that 0 is no timeout?

xiang90 commented on Documentation/configuration.md:
```diff
@@ -105,6 +105,27 @@ To start etcd automatically using custom settings at startup in Linux, using a [
+
+##### -proxy-read-timeout
++ Time (in milliseconds) for a read to timeout.
++ Don't change this value if you use watches because they are using long polling requests.
```
same here :)

xiang90 commented on proxy/director.go:
```diff
@@ -22,24 +22,17 @@ import (
 
-const (
-	// amount of time an endpoint will be held in a failed
-	// state before being reconsidered for proxied requests
```
I think we should still keep the second sentence. (move the whole comment to configuration.md).

xiang90 commented on proxy/director.go:
```diff
@@ -22,24 +22,17 @@ import (
-		uf: urlsFunc,
+		uf:              urlsFunc,
+		failureWait:     failureWait,
+		refreshInterval: refreshInterval,
```
Where is `refreshInterval` in director struct got used? Can we remove it if it is not used anywhere?

xiang90 commented:
@webner This change LGTM.

Thanks for the contribution! Have you tested the default configuration? I hope we can test it for a while in a high load env (have both watch, read, write) before merging it.

webner commented:
I'm still testing the default configuration in our development cluster. So far it's looking good. Restarting a etcd node in our 3-node etcd cluster has now less impact on the worker nodes which are using an etcd proxy.

xiang90 commented:
Fix #2862 (fix the original reference)

xiang90 commented:
@webner Let me know when this is ready

xiang90 commented:
@webner Any update?

webner commented:
Looks good so far. The proxy is running stable in our development cluster (~20 machines) and I had no issues with the default settings.

xiang90 commented:
@webner gofmt checks failed. Can you gofmt the proxy/direct.go code?
